File: src/AddressSync/Helper/Mapper/ClubMapper.vb
Imports System.Runtime.Remoting.Messaging

Imports Newtonsoft.Json.Linq

Imports pxBook



Public Class ClubMapper : Inherits Mapper(Of pxKommunikation.pxAdressen, JObject)
    Public Shared BooleanToString As New Func(Of Object, Object)(Function(obj)
                                                                     Return If(CType(obj, Boolean), "True", "False")
                                                                 End Function)
    Public Shared StringToBoolean As New Func(Of Object, Object)(Function(obj)
                                                                     Return If(CType(obj, String).ToLower = "true", True, False)
                                                                 End Function)


    '**************************************************************************************MappingProperties für ClubRelated Werte (inkl. MemberKey = AdressNr in Proffix)
    Public Overloads Shared ReadOnly Property MappingProperies As List(Of IMappingProperty)
        Get
            Return New List(Of IMappingProperty)({
                                              New MappingProperty("IsGliderPilot", "Z_Segelflugpilot", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderPilot", "Z_Segelflugpilot")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderPilot", "Z_Segelflugpilot"))),
                                              New MappingProperty("IsGliderInstructor", "Z_Segelfluglehrer", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderInstructor", "Z_Segelfluglehrer")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderInstructor", "Z_Segelfluglehrer"))),
                                              New MappingProperty("IsGliderTrainee", "Z_Segelflugschueler", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderTrainee", "Z_Segelflugschueler")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderTrainee", "Z_Segelflugschueler"))),
                                              New MappingProperty("IsMotorPilot", "Z_Motorflugpilot", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsMotorPilot", "Z_Motorflugpilot")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsMotorPilot", "Z_Motorflugpilot"))),
                                              New MappingProperty("IsPassenger", "Z_Passagier", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsPassenger", "Z_Passagier")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsPassenger", "Z_Passagier"))),
                                              New MappingProperty("IsTowPilot", "Z_Schleppilot", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsTowPilot", "Z_Schleppilot")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsTowPilot", "Z_Schleppilot"))),
                                              New MappingProperty("IsWinchOperator", "Z_Windenfuehrer", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsWinchOperator", "Z_Windenfuehrer")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsWinchOperator", "Z_Windenfuehrer"))),
                                              New MappingProperty("ReceiveFlightReports", "Z_erhaeltFlugreport", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceiveFlightReports", "Z_erhaeltFlugreport")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceiveFlightReports", "Z_erhaeltFlugreport"))),
                                              New MappingProperty("IsMotorInstructor", "Z_Motorfluglehrer", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsMotorInstructor", "Z_Motorfluglehrer")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsMotorInstructor", "Z_Motorfluglehrer"))),
                                              New MappingProperty("ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung"))),
                                              New MappingProperty("ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung", Nothing, Nothing, Nothing, Nothing,
                                                                    New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung"))),
                                              New MappingProperty("MemberStateId", "Z_MemberStateId", Nothing, Nothing, Nothing, Nothing,
                                                                   New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_ProffixZF(target, name, "MemberStateId", "Z_MemberStateId")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "MemberStateId", "Z_MemberStateId"))),
                                              New MappingProperty("MemberNumber", "AdressNr", Nothing, Nothing, Nothing, Nothing,
                                                                   New Func(Of Object, String, Object)(Function(target, name) GetValueFnc_JSONClubRel_Proffixdirekt(target, name, "MemberNumber", "AdressNr")),
                                                                    New Func(Of Object, String, Object, Object)(Function(target, name, value) SetValueFnc_JSONClubRel_Proffixdirekt(target, name, value, "MemberNumber", "AdressNr")))
 })

        End Get
    End Property

    '***********************************************************************Werte holen und einfügen*********************************************************************************
    ' für IsGliderPilot... (FLS ClubRelated <--> Proffix Zusatzfeld)
    Public Shared GetValueFnc_JSONClubRel_ProffixZF As New Func(Of Object, String, String, String, Object)(Function(target, name, serviceField, proffixField)
                                                                                                               Try
                                                                                                                   ' aus JSON ClubRel lesen
                                                                                                                   If target.GetType = GetType(JObject) Then
                                                                                                                       Return CType(target, JObject)("ClubRelatedPersonDetails")(serviceField)
                                                                                                                       ' aus Proffix Zusatzfeld lesen
                                                                                                                   ElseIf target.GetType = GetType(pxKommunikation.pxAdressen) Then
                                                                                                                       Return ProffixHelper.GetZusatzFelder(CType(target, pxKommunikation.pxAdressen), proffixField)
                                                                                                                   End If
                                                                                                                   Return target
                                                                                                               Catch ex As Exception
                                                                                                                   Throw New Exception("Fehler beim Auslesen von " + name)
                                                                                                                   Return Nothing
                                                                                                               End Try
                                                                                                           End Function)

    ' für MemberKey (FLS ClubRelated <--> Proffix Feld direkt)
    Public Shared GetValueFnc_JSONClubRel_Proffixdirekt As New Func(Of Object, String, String, String, Object)(Function(target, name, serviceField, proffixField)
                                                                                                                   Try
                                                                                                                       ' aus JSON ClubRel lesen
                                                                                                                       If target.GetType = GetType(JObject) Then
                                                                                                                           Return CType(target, JObject)("ClubRelatedPersonDetails")(serviceField)
                                                                                                                           ' direkt aus Proffix-Feld lesen
                                                                                                                       ElseIf target.GetType = GetType(pxKommunikation.pxAdressen) Then
                                                                                                                           Return target.GetType().GetField(name).GetValue(target)
                                                                                                                       End If
                                                                                                                       Return target
                                                                                                                   Catch ex As Exception
                                                                                                                       Throw New Exception("Fehler beim Auslesen von " + name)
                                                                                                                       Return Nothing
                                                                                                                   End Try
                                                                                                               End Function)


    ' für IsGliderPilot... (FLS ClubRelated <--> Proffix Zusatzfeld)
    Public Shared SetValueFnc_JSONClubRel_ProffixZF As New Func(Of Object, String, Object, String, String, Object)(Function(target, name, value, serviceField, proffixField)
                                                                                                                       ' in JSON einfügen (update --> clubrel vorhaden) (create --> clubrel als eigenes JObject erstellen)
                                                                                                                       If target.GetType = GetType(JObject) Then
                                                                                                                           ' 1/0 in boolean umwandeln und als Boolean in JObject schreiben
                                                                                                                           If value.ToString = "1" Or value.ToString = "0" Then
                                                                                                                               Dim JValue As JValue = CType(False, JValue)
                                                                                                                               If value.ToString = "1" Then
                                                                                                                                   JValue = CType(True, JValue)
                                                                                                                               ElseIf value.ToString = "0" Then
                                                                                                                                   JValue = CType(False, JValue)
                                                                                                                               End If

                                                                                                                               ' enthält das JObject den Text ClubRelatedPersonDetails?
                                                                                                                               If target.ToString.Contains("ClubRelatedPersonDetails") Then
                                                                                                                                   ' wenn ja (update von bestehender Adresse in FLS) --> unter clubrel einfügen
                                                                                                                                   CType(target, JObject)("ClubRelatedPersonDetails")(serviceField) = JValue
                                                                                                                               Else
                                                                                                                                   ' wenn nein (= create von neuer Adresse in FLS) --> clubrel als eigenes JObject erstellen und danach als ganzes in person einfügen
                                                                                                                                   CType(target, JObject)(serviceField) = JValue
                                                                                                                               End If
                                                                                                                               Return target
                                                                                                                           End If

                                                                                                                           ' ein String soll eingefügt werden
                                                                                                                           ' enthält das JObject den Text ClubRelatedPersonDetails?
                                                                                                                           If target.ToString.Contains("ClubRelatedPersonDetails") Then
                                                                                                                               ' wenn ja (update von bestehender Adresse in FLS) --> unter clubrel einfügen
                                                                                                                               CType(target, JObject)("ClubRelatedPersonDetails")(serviceField) = value.ToString
                                                                                                                           Else
                                                                                                                               ' wenn nein (= create von neuer Adresse in FLS) --> clubrel als eigenes JObject erstellen und danach als ganzes in person einfügen
                                                                                                                               CType(target, JObject)(serviceField) = value.ToString
                                                                                                                           End If

                                                                                                                           ' in pxAdresse Zusatzfeld einfügen
                                                                                                                       ElseIf target.GetType = GetType(pxKommunikation.pxAdressen) Then
                                                                                                                           Return ProffixHelper.SetZusatzFelder(
                                                                                                                                                                CType(target, pxKommunikation.pxAdressen),
                                                                                                                                                                proffixField,
                                                                                                                                                                proffixField,
                                                                                                                                                                "",
                                                                                                                                                                String.Empty,
                                                                                                                                                                If(value IsNot Nothing, value.ToString, String.Empty))
                                                                                                                       End If
                                                                                                                       Return target
                                                                                                                   End Function)


    ' für MemberNumber (FLS ClubRelated <--> Proffix Feld direkt)
    Public Shared SetValueFnc_JSONClubRel_Proffixdirekt As New Func(Of Object, String, Object, String, String, Object)(Function(target, name, value, serviceField, proffixField)
                                                                                                                           'MemberNumber bei Update und neu erstellen in FLS einfügen
                                                                                                                           If target.GetType = GetType(JObject) Then
                                                                                                                               ' wird update (clubrel bereits vorhanden) oder create (clubrel als eigenes JObject) gemacht
                                                                                                                               If target.ToString.Contains("ClubRelatedPersonDetails") Then
                                                                                                                                   ' wenn update --> clubrel bereits in FLS vorhanden --> unter clubrel einfügen
                                                                                                                                   CType(target, JObject)("ClubRelatedPersonDetails")(serviceField) = value.ToString
                                                                                                                               Else
                                                                                                                                   ' wenn create --> clubrel als eigenes JObject erstellt --> direkt einfügen (wird später als ganzes in person eingefügt)
                                                                                                                                   CType(target, JObject)(serviceField) = value.ToString
                                                                                                                               End If
                                                                                                                           ElseIf target.GetType = GetType(pxKommunikation.pxAdressen) Then
                                                                                                                               ' AdressNr muss nie selber gesetzt werden, da sie in Proffix erstellt wird (create), oder bereits vorhanden ist (update)
                                                                                                                           End If
                                                                                                                           Return target
                                                                                                                       End Function)


    ' '**************************************************************************************JObject clubPers in JObject pers einfügen, gibt zusammengefügtes JSON zurück
    Public Function completePersWithclubPers(ByVal target As JObject, ByRef innerJObject As JObject) As JObject
        CType(target, JObject)("ClubRelatedPersonDetails") = innerJObject
        Return target
    End Function

    '********************************************************************************Mapp Demapp*************************************************************************************************
    'Public Overloads Function Mapp(ByRef source As pxKommunikation.pxAdressen, ByRef target As JObject) As JObject
    '    'SetMemberKey(target, source)
    '    Return MyBase.Mapp(source, target)
    'End Function

    'Public Overloads Function DeMapp(ByRef source As JObject, ByRef target As pxKommunikation.pxAdressen) As pxKommunikation.pxAdressen
    '    Return MyBase.DeMapp(target, source)
    'End Function

End Class

File: src/AddressSync/Helper/Mapper/ClubMapper.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualBasic.CompilerServices;
using Newtonsoft.Json.Linq;
using pxBook;

namespace FlsGliderSync
{
    public class ClubMapper : Mapper<pxKommunikation.pxAdressen, JObject>
    {
        public static Func<object, object> BooleanToString = new Func<object, object>(obj => Conversions.ToBoolean(obj) ? "True" : "False");
        public static Func<object, object> StringToBoolean = new Func<object, object>(obj => Conversions.ToString(obj).ToLower() == "true" ? true : false);


        // **************************************************************************************MappingProperties für ClubRelated Werte (inkl. MemberKey = AdressNr in Proffix)
        public new static List<IMappingProperty> MappingProperies
        {
            get
            {
                return new List<IMappingProperty>(new[] { new MappingProperty("IsGliderPilot", "Z_Segelflugpilot", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderPilot", "Z_Segelflugpilot")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderPilot", "Z_Segelflugpilot"))), new MappingProperty("IsGliderInstructor", "Z_Segelfluglehrer", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderInstructor", "Z_Segelfluglehrer")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderInstructor", "Z_Segelfluglehrer"))), new MappingProperty("IsGliderTrainee", "Z_Segelflugschueler", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsGliderTrainee", "Z_Segelflugschueler")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsGliderTrainee", "Z_Segelflugschueler"))), new MappingProperty("IsMotorPilot", "Z_Motorflugpilot", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsMotorPilot", "Z_Motorflugpilot")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsMotorPilot", "Z_Motorflugpilot"))), new MappingProperty("IsPassenger", "Z_Passagier", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsPassenger", "Z_Passagier")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsPassenger", "Z_Passagier"))), new MappingProperty("IsTowPilot", "Z_Schleppilot", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsTowPilot", "Z_Schleppilot")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsTowPilot", "Z_Schleppilot"))), new MappingProperty("IsWinchOperator", "Z_Windenfuehrer", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsWinchOperator", "Z_Windenfuehrer")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsWinchOperator", "Z_Windenfuehrer"))), new MappingProperty("ReceiveFlightReports", "Z_erhaeltFlugreport", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceiveFlightReports", "Z_erhaeltFlugreport")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceiveFlightReports", "Z_erhaeltFlugreport"))), new MappingProperty("IsMotorInstructor", "Z_Motorfluglehrer", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "IsMotorInstructor", "Z_Motorfluglehrer")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "IsMotorInstructor", "Z_Motorfluglehrer"))), new MappingProperty("ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceiveAircraftReservationNotifications", "Z_erhaeltReservationsmeldung"))), new MappingProperty("ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "ReceivePlanningDayRoleReminder", "Z_erhaeltPlanungserinnerung"))), new MappingProperty("MemberStateId", "Z_MemberStateId", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_ProffixZF(target, name, "MemberStateId", "Z_MemberStateId")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_ProffixZF(target, name, value, "MemberStateId", "Z_MemberStateId"))), new MappingProperty("MemberNumber", "AdressNr", null, null, null, null, new Func<object, string, object>((target, name) => GetValueFnc_JSONClubRel_Proffixdirekt(target, name, "MemberNumber", "AdressNr")), new Func<object, string, object, object>((target, name, value) => SetValueFnc_JSONClubRel_Proffixdirekt(target, name, value, "MemberNumber", "AdressNr"))) });
            }
        }

        // ***********************************************************************Werte holen und einfügen*********************************************************************************
        // für IsGliderPilot... (FLS ClubRelated <--> Proffix Zusatzfeld)
        // aus JSON ClubRel lesen
        // aus Proffix Zusatzfeld lesen
        public static Func<object, string, string, string, object> GetValueFnc_JSONClubRel_ProffixZF = new Func<object, string, string, string, object>((target, name, serviceField, proffixField) => { try { if (target.GetType() == typeof(JObject)) { return ((JObject)target)["ClubRelatedPersonDetails"][serviceField]; } else if (target.GetType() == typeof(pxKommunikation.pxAdressen)) { return ProffixHelper.GetZusatzFelder((pxKommunikation.pxAdressen)target, proffixField); } return target; } catch (Exception ex) { throw new Exception("Fehler beim Auslesen von " + name); return null; } });

        // für MemberKey (FLS ClubRelated <--> Proffix Feld direkt)
        // aus JSON ClubRel lesen
        // direkt aus Proffix-Feld lesen
        public static Func<object, string, string, string, object> GetValueFnc_JSONClubRel_Proffixdirekt = new Func<object, string, string, string, object>((target, name, serviceField, proffixField) => { try { if (target.GetType() == typeof(JObject)) { return ((JObject)target)["ClubRelatedPersonDetails"][serviceField]; } else if (target.GetType() == typeof(pxKommunikation.pxAdressen)) { return target.GetType().GetField(name).GetValue(target); } return target; } catch (Exception ex) { throw new Exception("Fehler beim Auslesen von " + name); return null; } });


        // für IsGliderPilot... (FLS ClubRelated <--> Proffix Zusatzfeld)
        public static Func<object, string, object, string, string, object> SetValueFnc_JSONClubRel_ProffixZF = new Func<object, string, object, string, string, object>((target, name, value, serviceField, proffixField) =>
            {
            // in JSON einfügen (update --> clubrel vorhaden) (create --> clubrel als eigenes JObject erstellen)
            if (target.GetType() == typeof(JObject))
                {
                // 1/0 in boolean umwandeln und als Boolean in JObject schreiben
                if (value.ToString() == "1" | value.ToString() == "0")
                    {
                        JValue JValue = (JValue)false;
                        if (value.ToString() == "1")
                        {
                            JValue = (JValue)true;
                        }
                        else if (value.ToString() == "0")
                        {
                            JValue = (JValue)false;
                        }

                    // enthält das JObject den Text ClubRelatedPersonDetails?
                    if (target.ToString().Contains("ClubRelatedPersonDetails"))
                        {
                        // wenn ja (update von bestehender Adresse in FLS) --> unter clubrel einfügen
                        ((JObject)target)["ClubRelatedPersonDetails"][serviceField] = JValue;
                        }
                        else
                        {
                        // wenn nein (= create von neuer Adresse in FLS) --> clubrel als eigenes JObject erstellen und danach als ganzes in person einfügen
                        ((JObject)target)[serviceField] = JValue;
                        }

                        return target;
                    }

                // ein String soll eingefügt werden
                // enthält das JObject den Text ClubRelatedPersonDetails?
                if (target.ToString().Contains("ClubRelatedPersonDetails"))
                    {
                    // wenn ja (update von bestehender Adresse in FLS) --> unter clubrel einfügen
                    ((JObject)target)["ClubRelatedPersonDetails"][serviceField] = value.ToString();
                    }
                    else
                    {
                    // wenn nein (= create von neuer Adresse in FLS) --> clubrel als eigenes JObject erstellen und danach als ganzes in person einfügen
                    ((JObject)target)[serviceField] = value.ToString();
                    }
                }

            // in pxAdresse Zusatzfeld einfügen
            else if (target.GetType() == typeof(pxKommunikation.pxAdressen))
                {
                    return ProffixHelper.SetZusatzFelder((pxKommunikation.pxAdressen)target, proffixField, proffixField, "", string.Empty, value is object ? value.ToString() : string.Empty);
                }

                return target;
            });


        // für MemberNumber (FLS ClubRelated <--> Proffix Feld direkt)
        public static Func<object, string, object, string, string, object> SetValueFnc_JSONClubRel_Proffixdirekt = new Func<object, string, object, string, string, object>((target, name, value, serviceField, proffixField) =>
            {
            // MemberNumber bei Update und neu erstellen in FLS einfügen
            if (target.GetType() == typeof(JObject))
                {
                // wird update (clubrel bereits vorhanden) oder create (clubrel als eigenes JObject) gemacht
                if (target.ToString().Contains("ClubRelatedPersonDetails"))
                    {
                    // wenn update --> clubrel bereits in FLS vorhanden --> unter clubrel einfügen
                    ((JObject)target)["ClubRelatedPersonDetails"][serviceField] = value.ToString();
                    }
                    else
                    {
                    // wenn create --> clubrel als eigenes JObject erstellt --> direkt einfügen (wird später als ganzes in person eingefügt)
                    ((JObject)target)[serviceField] = value.ToString();
                    }
                }
                else if (target.GetType() == typeof(pxKommunikation.pxAdressen))
                {
                // AdressNr muss nie selber gesetzt werden, da sie in Proffix erstellt wird (create), oder bereits vorhanden ist (update)
            }

                return target;
            });


        // '**************************************************************************************JObject clubPers in JObject pers einfügen, gibt zusammengefügtes JSON zurück
        public JObject completePersWithclubPers(JObject target, ref JObject innerJObject)
        {
            target["ClubRelatedPersonDetails"] = innerJObject;
            return target;
        }

        // ********************************************************************************Mapp Demapp*************************************************************************************************
        // Public Overloads Function Mapp(ByRef source As pxKommunikation.pxAdressen, ByRef target As JObject) As JObject
        // 'SetMemberKey(target, source)
        // Return MyBase.Mapp(source, target)
        // End Function

        // Public Overloads Function DeMapp(ByRef source As JObject, ByRef target As pxKommunikation.pxAdressen) As pxKommunikation.pxAdressen
        // Return MyBase.DeMapp(target, source)
        // End Function

    }
}